week4/session_9_testing_homework.py

```diff
@@ -0,0 +1,74 @@
+from unittest.mock import patch
+import pytest
+from week4.exception_handling_atm_machine import check_pin_code
+from week4.exception_handling_atm_machine import validate_pin
+from week4.exception_handling_atm_machine import validate_withdrawal_amount
+from week4.exception_handling_atm_machine import withdrawal
+
+"""
+Write at least 5 unit tests to test the functions used in the atm machine program
+
+If input is not an integer, raise an exception.
+If input is not the correct number, raise an exception.
+If input is the correct number, continue.
+If pin is entered incorrectly more than 3 times, raise an exception.
+If withdrawal input is not an integer, raise an exception.
+If withdrawal is more than available balance, raise an exception and quit program.
+If withdrawal amount is less than available balance, continue.
+"""
+
+@patch('builtins.input', lambda *args: 'asdf')
+def test_check_pin_code_non_integer_input():
+    with pytest.raises(Exception) as exception_info:
+        check_pin_code()
+    assert "too many times" in str(exception_info.value)
+
+
+def test_validate_pin_raises_exception():
+    with pytest.raises(Exception):
+        pin = 3456
+        validate_pin(pin)
+
+
+def test_validate_pin_passes():
+    try:
+        pin = 1234
+        validate_pin(pin)
+    except:
+        pytest.fail("The pin number is incorrect")
+
+
+@patch('builtins.input', lambda *args: '12345')
+def test_check_pin_code_number_of_entries():
+    with pytest.raises(Exception) as exception_info:
+        check_pin_code()
+    assert "too many times" in str(exception_info.value)
+
+
+@patch('builtins.input', lambda *args: 'asdf')
+def test_withdrawal_non_integer_input(capsys):
+    withdrawal(40)
+    captured = capsys.readouterr()
+    assert "Please enter a number." in str(captured.out)
+
+
+def test_validate_withdrawal_amount_raises_exception():
+    with pytest.raises(Exception):
+        withdrawal_amount = 105
+        account_balance = 100
+        validate_withdrawal_amount(withdrawal_amount, account_balance)
+
+
+def test_validate_withdrawal_amount_passes():
+    try:
+        withdrawal_amount = 50
+        account_balance = 100
+        validate_withdrawal_amount(withdrawal_amount, account_balance)
+    except:
+        pytest.fail("The withdrawal amount is more than account balance")
+
+
```

Comment:
When running your python file, it runs the atm program again. This is because your test file isn't split into classes. When you split into classes, you pass in the unittest.TestCase which creates a test case. 
You are also missing the final block `unittest.main()` to run the tests and so I get :
```
Ran 0 tests in 0.000s

OK
```

Reply:
Hi Shivangi, 

I wrote my code using pytest rather than unittest. 

To run the tests `pip install pytest` and then run in the terminal using `pytest week4/session_9_testing_homework.py`

Thanks!

Reply:
Noted, Jo. The tests check out. Great work!

Reply:
I did notice you were importing `pytest`, however also saw you're importing `unittest`, so was running the tests using `unittest`. In the future, if you're using something different, you can explicitly declare it at the top of your file as a comment or using doc strings.

Reply:
Apologies! Have noted for future.

Reply:
No worries at all! Great job once again!